File: script/benchmark/benchmark.js
// 这个脚本跟lua脚本的benchmark.lua功能类似,主用来跑在v8或者quickjs上面对比验证性能
// v8可以这么测： node --expose-gc benchmark.js
// quickjs可以这么测： qjs benchmark.js
// quickjs没有jit比较拉胯，可以使用AOT
//      qjsc -o benchmark.c -c benchmark.js     # 生成 C 源码，包含字节码
//      gcc -o hello hello.c -lquickjs   # 编译为可执行文件
function benchmark(name, func) {
    const start = performance.now();
    func();
    const duration = (performance.now() - start) / 1000;
    console.log(`[${name.padEnd(20)}] ${duration.toFixed(3)} seconds`);
}

// 基准1：整数加法
function intArith() {
    let sum = 0;
    for (let i = 1; i <= 1e8; i++) {
        sum += i;
    }
}

// 基准2：浮点运算
function floatArith() {
    let sum = 0.0;
    for (let i = 1; i <= 1e7; i++) {
        sum += Math.sin(i) * Math.cos(i);
    }
}

// 基准3：表操作（对象/数组）
function tableOps() {
    const arr = new Array(1e6);
    for (let i = 0; i < 1e6; i++) {
        arr[i] = i;
    }
    let sum = 0;
    for (let i = 0; i < 1e6; i++) {
        sum += arr[i];
    }
}

// 基准4：递归
function recurse(n) {
    if (n === 0) return 0;
    return 1 + recurse(n - 1);
}

// 跑测试
benchmark("Integer Arithmetic", intArith);
benchmark("Float Arithmetic", floatArith);
benchmark("Table Operations", tableOps);
benchmark("Recursion (Depth 5000)", () => recurse(5000));

File: script/benchmark/benchmark.lua
local function benchmark(name, func)
    local start = os.clock()
    func()
    local duration = os.clock() - start
    print(string.format("[%-20s]  %.3f seconds", name, duration))
end

-- 基准1：整数加法
local function int_arith()
    local sum = 0
    for i = 1, 1e8 do
        sum = sum + i
    end
end

-- 基准2：浮点运算
local function float_arith()
    local sum = 0.0
    for i = 1, 1e7 do
        sum = sum + math.sin(i) * math.cos(i)
    end
end

-- 基准3：表操作
local function table_ops()
    local t = {}
    for i = 1, 1e6 do
        t[i] = i
    end
    local sum = 0
    for i = 1, 1e6 do
        sum = sum + t[i]
    end
end

-- 基准4：递归
local function recurse(n)
    if n == 0 then return 0 end
    return 1 + recurse(n - 1)
end

-- 跑测试
benchmark("Integer Arithmetic", int_arith)
benchmark("Float Arithmetic", float_arith)
benchmark("Table Operations", table_ops)
benchmark("Recursion (Depth 5000)", function() recurse(5000) end)
